### ISSUE TITLE
strictNullChecks support in v4.0 is broken

### ISSUE THREAD

**Issue by IgorMinar** [angular/angular#15432]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
Originally we implemented `strictNullChecks` support by hand-modifying the public api surface and testing apps against that with `strictNullChecks` tsc option on. We though we got a good coverage and fixed all the issues, but during the RC we started noticing that we missed quite a few places.

Eventually @mhevery went off to implement this properly by actually making the Angular codebase compile with `strictNullChecks` on without hand-tweaking the API. This work (#15405) surfaced a lot more issues with our public api surface that affects apps that do turn on `strictNullChecks`.

To make matters worse, we realized that if we were to allow apps to turn on `strictNullChecks` with these bugs in our public api surface, we wouldn't be able to fix the issues without breaking the apps that do turn on `strictNullChecks`.

**Expected behavior**
Correct `strictNullChecks` support.

**Proposed plan**
- Prevent applications from turning on `strictNullChecks` for now via a bogus symbol in our api that fails the strict null check
- Release 4.0 without the `strictNullChecks` support
- Merge @mhevery #15405 and any other follow up PRs related to this
- Release the `strictNullChecks` support in 4.1 (due in April, approximately one month from now)

**Comment by peterkelly** [comment]
Why are you actively introducing code to prevent strictNullChecks, when it would otherwise work properly for at least a subset of cases?

In a project I am working on, we have successfully been using strictNullChecks since 1.6. During our migration to Angular 2, we had to have our build process apply some patches we created a small number of definition files in the API so that our code would continue to build, but other than that it has been working fine (including in combination with Ionic 2.0).

I can understand the desire to ensure that the public API definition files properly reflect whether or not certain functions can accept or return values that are null or undefined, and that introducing null or undefined to a type constitutes a breaking API change. However, given that there are already existing projects that are using Angular with strictNullChecks (e.g. because they had a large existing codebase which already worked with it, and added Angular 2+ as a dependency), these changes are already breaking anyway.

Perhaps it might be worthwhile distributing strictNullCheck-compatible definition files that can be used as an option instead? I'm tempted to do so as part of a separate npm package, but if this is likely to be resolved shortly then it's probably better for me to wait it out.

**Comment by mhevery** [comment]
Because the type information is broken, and once we fix it we would break a lot of people, hence we could not roll it out in v4.1, but have to wait until v5. So by preventing people, we are saying this never worked, and it becomes a feature of v4.1.

If you really insist on doing `strictNullChecks` in your code knowing that you will be broken in v4.1 than you can put ` "skipLibCheck": true,` in your `tsconfig.json` which would allow you to turn on `"strictNullChecks": true,` without braking your build.

**Comment by aluanhaddad** [comment]
@mhevery As you know, this disables checking of all `.d.ts` files. This includes hand authored application level declarations. Although these files can generally be simply changed to `.ts`, it is not always the case.

Regardless, disabling checking of all libraries is not an acceptable workaround.

It is toxic.

**Comment by peterkelly** [comment]
I think I'm just going to stick to Angular 2

**Comment by mhevery** [comment]
BTW, enabling string null checks is on top of my priority list, so this should be fixed soon.

**Comment by sachinjkarche** [comment]
Hi all i am new to angular 2, i am learning. and today got to know angular 4 already came. Is there any angular 4 template for visual studio 2015 like angular 2?

**Comment by Delagen** [comment]
As for me compiler error only on forms and http components

**Comment by mhevery** [comment]
Support landed in v4.1

**Comment by CSchulz** [comment]
There is still an open bug with forms #16357

**Comment by Delagen** [comment]
@CSchulz see my comment, not only

**Comment by peter** [comment]
As of 10:th of October 2017 I cannot compile the latest released version of Angular (4.4.4) with latest TypeScript (2.5.3) and scrictNullChecks. Failures:

Error at node_modules/@angular/core/src/change_detection/differs/default_iterable_differ.d.ts:2:22: Class 'DefaultIterableDifferFactory' incorrectly implements interface 'IterableDifferFactory'.
  Types of property 'create' are incompatible.
    Type '<V>(trackByFn?: TrackByFunction<V> | undefined) => DefaultIterableDiffer<V>' is not assignable to type '{ <V>(trackByFn?: TrackByFunction<V> | undefined): IterableDiffer<V>; <V>(_cdr?: ChangeDetectorRe...'.

Error at node_modules/@angular/core/src/change_detection/differs/default_iterable_differ.d.ts:10:22: Class 'DefaultIterableDiffer<V>' incorrectly implements interface 'IterableChanges<V>'.
  Types of property 'forEachOperation' are incompatible.
    Type '(fn: (item: IterableChangeRecord<V>, previousIndex: number | null, currentIndex: number | null) =...' is not assignable to type '(fn: (record: Iterabl

Error at node_modules/@angular/core/src/change_detection/differs/default_iterable_differ.d.ts:10:22: Class 'DefaultIterableDiffer<V>' incorrectly implements interface 'IterableDiffer<V>'.
  Types of property 'diff' are incompatible.
